doc/designs/quic-design/connection-id-cache.md

```diff
@@ -0,0 +1,257 @@
+
+/* New functions */
+int ossl_quic_conn_id_set(QUIC_CONN_ID *cid, unsigned char *id,
+                          unsigned int id_len);
+
+int ossl_quic_conn_id_generate(QUIC_CONN_ID *cid);
+```
+
+### Remote Connection ID APIs
+
+```c
+typedef struct quic_remote_conn_id_cache_st QUIC_REMOTE_CONN_ID_CACHE;
+
+/*
+ * Allocate and free a remote connection ID cache
+ */
+QUIC_REMOTE_CONN_ID_CACHE *ossl_quic_remote_conn_id_cache_new_conn(
+        unsigned int id_limit   /* [active_connection_id_limit] */
+    );
+void ossl_quic_remote_conn_id_cache_free(QUIC_REMOTE_CONN_ID_CACHE *cache);
+
+/*
+ * Add a remote connection ID to the cache
+ */
+int ossl_quic_cache_add_remote_conn_id(QUIC_REMOTE_CONN_ID_CACHE *cache,
+                                       const QUIC_CONNECTION *conn,
+                                       const unsigned char *conn_id,
+                                       size_t conn_id_len,
+                                       uint64_t seq_no);
+
+/*
+ * Query a remote connection for a connection ID.
+ * Each connection can have multiple connection IDs associated with different
+ * routes.  This function returns one of these in a non-specified manner.
+ */
+int ossl_quic_conn_id_cache_get0_remote_conn_id(
+        const QUIC_REMOTE_CONN_ID_CACHE *cache,
+        const QUIC_CONNECTION *conn, QUIC_CONN_ID **cid);
+
+/*
+ * Retire remote connection IDs up to and including the one determined by the
+ * sequence number.
+ */
+int ossl_quic_conn_id_cache_retire_remote_conn_ids(
+        QUIC_REMOTE_CONN_ID_CACHE *cache, uint64_t seq_no);
+
+/*
+ * Remove remote connection IDs up to and including the one determined by the
+ * sequence number.
+ */
+int ossl_quic_conn_id_cache_retire_remove_conn_ids(
+        QUIC_REMOTE_CONN_ID_CACHE *cache, uint64_t seq_no);
+```
+
+### Local Connection ID APIs
+
+```c
+typedef struct quic_local_conn_id_cache_st QUIC_LOCAL_CONN_ID_CACHE;
+
+/*
+ * Allocate and free a local connection ID cache
+ */
+QUIC_LOCAL_CONN_ID_CACHE *ossl_quic_local_conn_id_cache_new_conn(void);
+void ossl_quic_local_conn_id_cache_free(QUIC_LOCAL_CONN_ID_CACHE *cache);
+
+/*
+ * Generate a new random local connection ID and associate it with a connection.
+ * For MVP this could just be a zero length ID.
+ */
+int ossl_quic_local_conn_id_cache_new_conn_id(QUIC_LOCAL_CONN_ID_CACHE *cache,
+                                              QUIC_CONNECTION *conn,
+                                              QUIC_CONN_ID **cid);
+
+/*
+ * Remove a local connection and all associated cached IDs
+ */
+int ossl_quic_remove_connection(QUIC_LOCAL_CONN_ID_CACHE *cache,
+                                const QUIC_CONNECTION *conn);
+
+/*
+ * Lookup a local connection by ID.
+ * Returns the connection or NULL if absent.
+ */
+QUIC_CONNECTION *ossl_quic_get0_local_conn(
+        const QUIC_LOCAL_CONN_ID_CACHE *cache,
+        const unsigned char *conn_id, size_t conn_id_len);
```

Comment:
Why do we sometimes pass connection ids as a pointer to a `QUIC_CONN_ID` object (which just contains the pair of the connection id buffer and its length), and other times we pass the buffer and its length directly?

Reply:
The idea was to wrap the creation of the QUIC_CONN_ID objects inside these routines.  The caller can then pass pointers to packet buffers rather than extracting the data and passing fully fledged IDs.
This is the difference between:
```c
QUIC_CONN_ID id;

ossl_quic_conn_id_set(&id, buf_ptr, len);
ossl_quic_get0_local_conn(cache, &id);
```
and:
```c
ossl_quic_get0_local_conn(cache, buf_ptr, len);
```

I think the latter is the nicer option both from writing and understanding.  It also saves a copy and reduces memory/cache use but that's not a big deal here.